v2/datastream-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/spanner/ProcessInformationSchema.java

```diff
@@ -112,34 +126,70 @@ public void setup() throws Exception {
     @Teardown
     public void teardown() throws Exception {
       spannerAccessor.close();
+      shadowTableSpannerAccessor.close();
     }
 
     @ProcessElement
     public void processElement(ProcessContext c) {
+      // TODO: Add pgsql support/ cross dialect support.
+      Ddl mainDdl = getInformationSchemaAsDdl(spannerAccessor);
+      Ddl shadowTableDdl = getInformationSchemaAsDdl(shadowTableSpannerAccessor);
+
       if (shouldCreateShadowTables) {
-        createShadowTablesInSpanner(getInformationSchemaAsDdl());
+        createShadowTablesInSpanner(mainDdl, shadowTableDdl);
+        // Refresh shadow table DDL after creating new shadow tables
+        shadowTableDdl = getInformationSchemaAsDdl(shadowTableSpannerAccessor);
       }
-      c.output(getInformationSchemaAsDdl());
+      // Clean up DDLs to ensure proper separation of main and shadow tables. Note that this is only
+      // for data hygiene, the downstream methods using these can handle Ddl with excess tables as
+      // well.
+      mainDdl = cleanupDdl(mainDdl, shadowTablePrefix, /* isMainDdl= */ true);
+      shadowTableDdl = cleanupDdl(shadowTableDdl, shadowTablePrefix, /* isMainDdl= */ false);
+      c.output(MAIN_DDL_TAG, mainDdl);
+      c.output(SHADOW_TABLE_DDL_TAG, shadowTableDdl);
+    }
+
+    Ddl cleanupDdl(Ddl originalDdl, String shadowPrefix, boolean isMainDdl) {
```

Comment:
why is this required ?

Reply:
I have added a coment above on why its done. This is not a requirement but a data hygiene thing.